src/monitor.py

```diff
@@ -115,3 +119,136 @@ def read_recent_logs(self, job: JobDto, tail: int = 20) -> str:
+            ]
+        failure_rollouts_messages = [
+            f'deployment \"{resource_name}\" exceeded its progress deadline'
+        ]
+
+        cmd = f'kubectl rollout status deployment/{resource_name} --watch={watch} --namespace {K8S_NAMESPACE}'
+        rollout_status = shell_output(cmd)
+
+        if any(msg in rollout_status for msg in successful_rollouts_messages):
+            return True, rollout_status
+        if any(msg in rollout_status for msg in failure_rollouts_messages):
+            return False, rollout_status
+        return None, rollout_status
+
+
+    def check_k8s_events_for_errors(self, resource_name: str) -> tuple[bool | None, str]:
+        # TODO: A single "kubectl get events" command could be used to get all the events at once, but this is easier to reason about
+
+        # First, query relevant Deployment events based on resource_name, e.g. "job-adder-v-0-0-2"
+        base_cmd = f'kubectl get events ' + \
+            f'--namespace {K8S_NAMESPACE} ' + \
+            f'--sort-by=\'.metadata.creationTimestamp\' ' + \
+            f'-o json'
+        deployment_selectors = f'--field-selector involvedObject.kind=Deployment,involvedObject.name={resource_name}'
+        deployment_query = json.loads(shell_output(f'{base_cmd} {deployment_selectors}'))
+
+        # From the deployment query, we extract the relevant ReplicaSet names
+        replicaset_names = self.get_replicaset_names_from_deployment_query(deployment_query, resource_name)
+
+        # Query for all the ReplicaSets
+        replicaset_queries = [
+            json.loads(shell_output(f'{base_cmd} --field-selector involvedObject.kind=ReplicaSet,involvedObject.name={name}'))
+            for name in replicaset_names
+        ]
+
+        # From the ReplicaSet queries, we get the relevant Pod names
+        pod_names = self.get_pod_names_from_replicaset_queries(replicaset_queries, resource_name)
+
+        # Query for all the Pods
+        pod_queries = [
+            json.loads(shell_output(f'{base_cmd} --field-selector involvedObject.kind=Pod,involvedObject.name={pod_name}'))
+            for pod_name in pod_names
+        ]
+
+        # Then, we check the pod queries for errors.
+        # only look for reasons
+        errors_we_dont_want_in_messages = ['Insufficient cpu', 'Insufficient memory', 'Insufficient ephemeral-storage']
+        errors_we_dont_want_in_reasons = ['FailedScheduling', 'SchedulerError', 'CrashLoopBackOff']
+        for pod_query in pod_queries:
+            for event in pod_query.get('items', []):
+                for error in errors_we_dont_want_in_messages:
+                    if error in event['message'] and event['type'] == 'Warning':
+                        return False, event
+                for error in errors_we_dont_want_in_reasons:
+                    if error in event['reason'] and event['type'] == 'Warning':
+                        return False, event
```

Comment:
I think this check could also state that the pod have been started correctly and return early, according to this example:
![Screenshot from 2024-05-15 10-15-30](https://github.com/TheRacetrack/plugin-kubernetes-infrastructure/assets/45496492/50ee4252-7e33-4672-8748-c9d855d45fdd)

```suggestion
                        return False, event
                if event['reason'] == 'Started':
                    return True, event
```